File: 3351-maximize-happiness-of-selected-children/maximize-happiness-of-selected-children.cpp
class Solution {
public:
    long long maximumHappinessSum(vector<int>& happiness, int k) {
        // 09.05.24 POTD
        long long ans=0;
        int t=0;
        sort(happiness.begin(), happiness.end());
        
        for(int i=happiness.size()-1; i>=0, k>0; i--, k--){
            if(happiness[i]-t > 0){
                ans += happiness[i]-t;
            } else {
                break;
            }
            t++;
        }

        return ans;
    }
};

File: 3351-maximize-happiness-of-selected-children/maximize-happiness-of-selected-children.py
class Solution:
    def maximumHappinessSum(self, christmasJoy: List[int], gifts: int) -> int:
        # 25.12.25 potd
        christmasJoy.sort(reverse=True)
        
        totalJoy = 0
        for turn in range(gifts):
            currentJoy = christmasJoy[turn] - turn
            if currentJoy <= 0:
                break
            totalJoy += currentJoy
        
        return totalJoy